src/realm/sync/network/default_socket.cpp

```diff
@@ -569,38 +569,39 @@ void DefaultSocketProvider::stop(bool wait_for_stop)
     std::unique_lock<std::mutex> lock(m_mutex);
 
     // Do nothing if the thread is not started or running or stop has already been called
-    if (m_state == State::Starting || m_state == State::Started || m_state == State::Running) {
+    if (m_state == State::Starting || m_state == State::Running) {
         m_logger_ptr->trace("Default event loop: stop()");
-        do_state_update(State::Stopping);
+        do_state_update(lock, State::Stopping);
         // Updating state to Stopping will free a start() if it is waiting for the thread to
         // start and may cause the thread to exit early before calling service.run()
         m_service.stop(); // Unblocks m_service.run()
     }
 
     // Wait until the thread is stopped (exited) if requested
-    if (wait_for_stop && m_state == State::Stopping) {
-        lock.unlock();
+    if (wait_for_stop) {
```

Comment:
I couldn't think of a way we'd have a state != State::Stopping || State::Stopped here, so we shouldn't need the extra check